src/sage/schemes/elliptic_curves/ell_number_field.py

```
@@ -4062,3 +4068,264 @@ def rational_points(self, **kwds):
+        + 1729396406547138/161051*x^13 + 2959141014912537/161051*x^12 + 4049663724749832/161051*x^11
+        + 4450611572214360/161051*x^10 + 3931865527421272/161051*x^9 + 2786855392905248/161051*x^8
+        + 1576387623537152/161051*x^7 + 704811439294144/161051*x^6 + 22297037611520/14641*x^5 + 5896572491264/14641*x^4
+        + 12558313746944/161051*x^3 + 152158103552/14641*x^2 + 136821293056/161051*x + 5155295232/161051
+
+    TESTS::
+
+        sage: _modular_polynomial('banana', 7, j)
+        Traceback (most recent call last):
+        ...
+        ValueError: (group, p) must be one of ...
+        sage: _modular_polynomial('borel', 101, j)
+        Traceback (most recent call last):
+        ...
+        ValueError: (group, p) must be one of ...
+    """
+    valid_parameters = [("borel", 3), ("borel", 5), ("borel", 7),
+                       ("split", 3), ("split", 5), ("split", 7),
+                       ("nonsplit", 5), ("nonsplit", 7)]
+    if (group, p) not in valid_parameters:
+        raise ValueError("(group, p) must be one of {}".format(valid_parameters))
+
+    from sage.rings.polynomial.polynomial_ring import polygen
+    x = polygen(j.parent())
+    if group == "borel":
+        from sage.schemes.elliptic_curves.isogeny_small_degree import Fricke_polynomial
+        return Fricke_polynomial(p)(x)-j*x
+    if group == "split":
+        if p == 3:
+            return (x-9)**3*(x+3)**3-j*x**3
+        if p == 5:
+            return ((x**2-5)*(x**2+5*x+10)*(x+5))**3-j*(x**2+5*x+5)**5
+        if p == 7:
+            return ((x**2-5*x+8)*(x**2-5*x+1)*(x**4-5*x**3+8*x**2-7*x+7)*(x+1))**3*x-j*(x**3-4*x**2+3*x+1)**7
+    if group == "nonsplit":
+        if p == 5:
+            return 5**4*(2*x+1)*(x+1)**3*(6*x**2+21*x+19)**3-j*(x**2+x-1)**5
+        if p == 7:
+            return ((4*x**2+5*x+2)*(x**2+3*x+4)*(x**2+10*x+4)*(3*x+1))**3-j*(x**3+x**2-2*x-1)**7
+
+    raise RuntimeError # cannot happen -- the assertion would have failed.
```

so this line is useless and should be removed

---

I find there is some utility in this as it would catch bugs (or at least make it easier to see where they are coming from) by raising a specific type of exception rather that silently returning `None`.

---

but this line cannot be reached at all, I think

---

That's the point. It should never be reached. I am treating it like an `assert` statement; something that should never be false/reached. Actually, would that be okay with you: `assert False`?

---

I am sorry, but I do not see the point. We are first checking the input parameters to belong to an explicit list. Then, for each member of the list, the code path ends by a return. So this *proves* that the final line can *never* happen.

---

You're exactly right that this code is _currently_ bug-free. Yet, suppose I change this method, but I introduce a bug for some case. Then I run this code, and without this `raise`, the code silently returns a wrong answer of `None`. Now this is highly likely to fail, but somewhere else. However, the failure is not guaranteed (no reasonable person should trap a `RuntimeError` or an `AssertionError`), which means a wrong answer could actually be returned. Thus, this one extra line makes the code more robust against future errors that could be introduced.

---

ok, ok. I do not agree, but will not oppose any longer, for doc and code inside a hidden method. Good for me